File: raspPi/src/nodes/MirrorNode.cpp
#include "MirrorNode.hpp"

Attribute pathAttrib("path");


bool MirrorNode::LinkMirror()
{
    BaseNode* pMirror = FindNode(mirrorPath, false, false);
    SetMirrorSource(pMirror);
    return IsLinked();
}

void MirrorNode::SetMirrorSource(BaseNode *pSource)
{
    if (pSource == nullptr)
    {
        if(pMirrorSource != nullptr) 
        {   // Remove old link.
            pMirrorSource->RemoveMirror(this);
            pMirrorSource = nullptr;
        }
        return;
    }

    if(pMirrorSource != nullptr &&  pMirrorSource != pSource)
    {  // Remove old link.
        pMirrorSource->RemoveMirror(this);
    }

    // Link to new source.
    pMirrorSource = pSource; 
    pMirrorSource->AddMirror(this);
}

void MirrorNode::SetMirrorPath(const char* pPath)
{
    mirrorPath = pPath;
    SetAttributeChanged(pathAttrib.GetID());
}

bool MirrorNode::SetAttribute(attribID_t attribID, const char* pAttributeValue)
{
    if (pathAttrib.GetID() == attribID)
    {
        SetMirrorPath(pAttributeValue);
        return true;
    }
    if (pMirrorSource == nullptr) return false;
    return pMirrorSource->SetAttribute(attribID, pAttributeValue);
}

bool MirrorNode::GetAttribute(attribID_t attribID, std::string &strAttributeValue) const
{
    if (pathAttrib.GetID() == attribID)
    {
        strAttributeValue = mirrorPath;
        return true;
    }if (pMirrorSource == nullptr) return false;
    return pMirrorSource->GetAttribute(attribID, strAttributeValue);
}

bool MirrorNode::IsAttributeChanged(attribID_t attribID) const
{
    if (pathAttrib.GetID() == attribID)
    {
        return BaseNode::IsAttributeChanged(attribID);
    }
    if (pMirrorSource == nullptr) return false;
    return pMirrorSource->IsAttributeChanged(attribID); 
}

bool MirrorNode::IsAttributeUsed(attribID_t attribID) const
{
    if (pathAttrib.GetID() == attribID)
    {
        return BaseNode::IsAttributeUsed(attribID);
    }
    if (pMirrorSource == nullptr) return false;
    return pMirrorSource->IsAttributeUsed(attribID);
}

File: ROBINinspect/nodes/MirrorNode.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ROBINinspect
{
    public class MirrorNode : BaseNode
    {
        public MirrorNode(String nodeName, String path = "")
            : base(nodeName)
        {
            mirrorPath = path;
        }

        private String mirrorPath;
        public String MirrorPath
        {
            get { return mirrorPath; }
            set
            {
                mirrorPath = value;
                SetAttributeChanged(AttributeTypes.path);
            }
        }

        private BaseNode mirrorSource;
        public BaseNode MirrorSource
        {
            get { return mirrorSource; }
        }

        public bool IsLinked() { return mirrorSource != null; }
        public bool LinkMirror()
        {
            BaseNode mirror = FindNode(mirrorPath, false, false);
            SetMirrorSource(mirror);
            return IsLinked();
        }

        void SetMirrorSource(BaseNode source)
        {
            if (source == null)
            {
                if(mirrorSource != null) 
                {   // Remove old link.
                    mirrorSource.RemoveMirror(this);
                    mirrorSource = null;
                }
                return;
            }

            if(mirrorSource != null &&  mirrorSource != source)
            {  // Remove old link.
                mirrorSource.RemoveMirror(this);
            }

            // Link to new source.
            mirrorSource = source; 
            mirrorSource.AddMirror(this);
        }

        public override bool GetAttribute(AttributeTypes attribID, ref String strAttributeValue)
        {
            if (attribID == AttributeTypes.path)
            {
                strAttributeValue = MirrorPath;
                return true;
            }
            if (mirrorSource == null)
            {
                return false;
            }
            else
            {
                return mirrorSource.GetAttribute(attribID,ref strAttributeValue);
            }
        }

        public override bool SetAttribute(AttributeTypes attribID, String strAttributeValue)
        {
            if (attribID == AttributeTypes.path)
            {
                MirrorPath = strAttributeValue;
                return true;
            }
            if (mirrorSource == null)
            {
                return false;
            }
            else
            {
                return mirrorSource.SetAttribute(attribID, strAttributeValue);
            }
        }

        public override bool IsAttributeChanged(AttributeTypes attribID)
        {
            if (attribID == AttributeTypes.path)
            {
                return base.IsAttributeChanged(attribID);
            }
            if (mirrorSource == null)
            {
                return false; 
            }
            else
            {
                return mirrorSource.IsAttributeChanged(attribID);
            }
        }

        public override bool IsAttributeUsed(AttributeTypes attribID)
        {
            if (attribID == AttributeTypes.path)
            {
                return base.IsAttributeUsed(attribID);
            }
            if (mirrorSource == null)
            {
                return false;
            }
            else
            {
                return mirrorSource.IsAttributeUsed(attribID);
            }
        }
    }
}
